edgedb/retry.py

```diff
@@ -0,0 +1,87 @@
+import asyncio
+import random
+
+from . import errors
+from . import transaction as _transaction
+
+
+DEFAULT_MAX_ITERATIONS = 3
+
+
+def default_backoff(attempt):
+    return (2 ** attempt) * 0.1 + random.randrange(100) * 0.1
+
+
+class AsyncIOIteration(_transaction.AsyncIOTransaction):
+    def __init__(self, retry, owner):
+        super().__init__(owner)
+        self.__retry = retry
+
+    async def start(self):
+        if not self._managed:
+            raise errors.InterfaceError(
+                "Only managed retriable transactions are supported. "
+                "Use `async with transaction:`"
+            )
+        # TODO(tailhook) if this is not the first iteration suppress
+        # `wait_until_available` timeout
+        await super().start()
+
+    async def __aexit__(self, extype, ex, tb):
+        try:
+            await super().__aexit__(extype, ex, tb)
+        except errors.ClientConnectionClosedError as err:
+            if ex is None:
+                # TODO(tailhook) should we figure out if err is *caused by*
+                # CancelledError and propagate the latter instead or retrying?
+
+                # If we were going to commit, retry
+                if self.__retry._retry(err):
+                    return True
+                else:
+                    # if retries are exhausted we need to propagate this error
+                    raise err
```

Comment:
IIRC you need to just `return False` here. The exception will be propagated by the interpreter. If you do it with a `raise` you'd likely skew its traceback.

Reply:
It's not the exception from body of the with. It's the exception we caught `from super().__exit__()`, it will not be reraised on return False.